template/interpolate/funcs.go

```diff
@@ -163,7 +163,15 @@ func funcGenUser(ctx *Context) interface{} {
 			return "", errors.New("test")
 		}
 
-		return ctx.UserVariables[k], nil
+		val, ok := ctx.UserVariables[k]
+		if ctx.EnableEnv {
+			// error and retry if we're interpolating UserVariables. But if
+			// we're elsewhere in the template, just return the empty string.
+			if !ok {
+				return "", errors.New(fmt.Sprintf("variable %s not set", k))
```

Comment:
This could be a typed error so that we can easily switch on it.

ex:

```
type ErrorUserVariableNotFound string

func (var ErrorUserVariableNotFound) Error() string {
       return fmt.Sprintf("variable %s not set", var)
}
```

Reply:
Turns out that template.exec eats all the errors and smushes them into an errors.ErrorString, so this path didn't work.

Reply:
ah I totally _expected_ it would pass it on in `ExecError.Err`. welp sorry ! 🙂